compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -486,6 +530,157 @@ abstract class GenJSCode extends plugins.PluginComponent
       classDefinition
     }
 
+    /** Generate an instance of an anonymous Scala.js defined class inline
+     *
+     *  @param sym Class to generate the instance of
+     *  @param args Arguments to the constructor
+     *  @param pos Position of the original New tree
+     */
+    def genAnonSJSDefinedNew(sym: Symbol, args: List[js.Tree],
+        pos: Position): js.Tree = {
+      assert(isScalaJSDefinedAnonJSClass(sym),
+          "Generating AnonSJSDefinedNew of non anonymous SJSDefined JS class")
+
+      // Find the ClassDef for this anonymous class
+      val classDef = consumeLazilyGeneratedAnonClass(sym)
+
+      // Generate a normal SJSDefinedJSClass
+      val origJsClass =
+        nestedGenerateClass(sym)(genScalaJSDefinedJSClass(classDef))
+
+      // Partition class members.
+      val staticMembers = ListBuffer.empty[js.Tree]
+      val classMembers = ListBuffer.empty[js.Tree]
+      var constructor: Option[js.MethodDef] = None
+
+      origJsClass.defs.foreach {
+        case fdef: js.FieldDef =>
+          classMembers += fdef
+
+        case mdef: js.MethodDef =>
+          mdef.name match {
+            case _: js.Ident =>
+              assert(mdef.static, "Non-static method in SJS defined JS class")
+              staticMembers += mdef
+
+            case js.StringLiteral(name) =>
+              assert(!mdef.static, "Exported static method")
+
+              if (name == "constructor") {
+                assert(constructor.isEmpty, "two ctors in class")
+                constructor = Some(mdef)
+              } else {
+                classMembers += mdef
+              }
+          }
+
+        case property: js.PropertyDef =>
+          classMembers += property
+
+        case tree =>
+          sys.error("Unexpected tree: " + tree)
+      }
+
+      // Only emit static members to class
+      generatedClasses += sym -> origJsClass.copy(
+          defs = staticMembers.toList)(
+          origJsClass.optimizerHints)(origJsClass.pos)
```

Comment:
You might also want to turn the class into a normal Scala class extending `Object`, here. So maybe it's easier to create a new `ClassDef` from scratch, rather than copying `origJsClass`.